lib/mongo_batch.rb

```diff
@@ -15,16 +15,20 @@ def batches
         processed_so_far = offset
 
         offset.step(by: batch_size, to: to - batch_size).each do |offset|
-          yielder << query.order_by(order_by).limit(batch_size).skip(offset)
+          yielder << with_order.limit(batch_size).skip(offset)
           processed_so_far += batch_size
         end
 
         if processed_so_far < to
           last_limit = to - processed_so_far
-          yielder << query.order(order_by).limit(last_limit).skip(processed_so_far)
+          yielder << with_order.limit(last_limit).skip(processed_so_far)
         end
       end
     end
+
+    def with_order
+      query.all.options.sort ? query.all : query.order_by(order_by)
```

Comment:
Do you have to specify `all` here or can you ask the `query` itself for its `options.sort`?

Playing in the console it seems that both `Mongoid::Criteria` and `Mongo::Collection::View` objects will respond to `options.sort` but the former will return `nil` and the latter `[]`.